Introduce line_to_axis_pos() for use in homeaxis()

AnHardt opened the pull request:
Introduce line_to_axis_pos(axis_codes axis, float where, float feed_rate = 0.0)
and use it in homeaxis instead of `do_blocking_move_to_axis_pos()`.

`do_blocking_move_to_axis_pos` was wrong because it performs subdivided, delta-corrected moves for x- and y-axis.

Hopefully a fix for https://github.com/MarlinFirmware/Marlin/pull/4338#issuecomment-233275637

@esenapaj Would you pleas be so kind to test?

ghost commented:
I've tested this PR just now, but this is danger than previous one... 

Video clip:
[![](http://img.youtube.com/vi/VcoUde3wNsI/0.jpg)](https://www.youtube.com/watch?v=VcoUde3wNsI)

A branch that it's used for test:
https://github.com/esenapaj/Marlin/tree/testes

AnHardt commented:
@esenapaj 
Sorry. I'll have an other look into the code now.

AnHardt commented:
@esenapaj 
Updated.
The insight that the first move in delta homing is like quick_home_xy(), but for 3 towers, may have helped.
As always while testing - finger on the reset button! Sorry - can't test it myself - no delta machine.

ghost commented:
I've tested commit 176099639288809c68ec7965f5586b8e4c9b67d2, but danger behavior is exactly the same as the 1fc3250267bf3531953f71db7c0f9e569913b49f...

jbrazio commented:
@AnHardt 

Context:
- `MIN_Z_HEIGHT_FOR_HOMING` is 5
- `M851 Z-1.50`

When I have the nozzle at Z5 (`X:71.00 Y:133.00 Z:6.50 E:0.00 Count X: 11360 Y:21280 Z:52000`) and issue a `G28` the tool is Z raised by 1.5mm and only then it starts searching for bed, when it finds the bed and finishes homing Z my Z position is 6.50 (`X:71.00 Y:133.00 Z:6.50 E:0.00 Count X: 11360 Y:21280 Z:52000`).

When I have the nozzle at Z6.50 (`X:71.00 Y:133.00 Z:6.50 E:0.00 Count X: 11360 Y:21280 Z:52000`) and issue a `G28` it will start imediatly searching for the bed, when it finds the bed and finishes homing Z my Z positions is 6.50 (`X:71.00 Y:133.00 Z:6.50 E:0.00 Count X: 11360 Y:21280 Z:52000`).

I would expect both of them to finish at Z5 and Z never being lift above 5 during the process.
I suspect the error is [around here](https://github.com/AnHardt/Marlin/blob/rep-cons-g28/Marlin/Marlin_main.cpp#L2981-L2992).

Issuing `G28X0` made an automatic X crash against the endstop, the same for `G28Y0`.

AnHardt commented:
@jbrazio 
"G28" does not home x and y for you?
I hope you described "G28 Z".
In this case ask @thinkyhead about 'do_probe_raise()' #4282.

The complete G28 sequence for a !DELTA, homing to z-min, with a probe should be:
- Raise to at least `MIN_Z_HEIGHT_FOR_HOMING`. For the first homing boot position is 0. (this is the part you have under suspicion )
- home x and y
- in case go to the safe home position
- deploy the probe (here 'do_probe_raise()' is used and the additional 1.5mm come in.)
- home z
- stow the probe (here 'do_probe_raise()' is used and the additional 1.5mm come in.)

jbrazio commented:
`G28` by itself homes with the quirk I described.
`G28 X` or `G28 Y` is completely broken with this PR.

PS: I use an induction probe.

AnHardt commented:
Wasted a day by looking into **My** code. The endstops are not initialized as expected. #4306 is under suspicion.

Workaround: 
Activate `#define ENDSTOPS_ALWAYS_ON_DEFAULT`

I'm too tired now to find the problem. Up since 24 hours after 3h of sleep.  More later.

ghost commented:
Thank you for your hard work.
But when I enabled `#define ENDSTOPS_ALWAYS_ON_DEFAULT`, I got more strange behavior.

Case 1:
<details>
<summary>

Newest RCBugFix, add PR #4342 (Introduce line_to_axis_pos() for use in homeaxis()), `#define ENDSTOPS_ALWAYS_ON_DEFAULT` is enabled</summary>



```
20:04:54.495 : Printer reset detected - initalizing
20:04:54.495 : start
20:04:54.500 : echoecho:Unknown command: ""
20:04:55.450 : Printer reset detected - initalizing
20:04:55.451 : start
20:04:55.451 : echo: External Reset
20:04:55.451 : Marlin 1.1.0-RCBugFix
20:04:55.451 : echo: Last Updated: 2016-04-27 12:00 | Author: (Micromake)
20:04:55.455 : Compiled: Jul 19 2016
20:04:55.455 : echo: Free Memory: 2604  PlannerBufferBytes: 1408
20:04:55.459 : echo:V24 stored settings retrieved (427 bytes)
20:04:55.459 : echo:Steps per unit:
20:04:55.459 : echo:  M92 X200.00 Y200.00 Z200.00 E476.55
20:04:55.463 : echo:Maximum feedrates (mm/s):
20:04:55.463 : echo:  M203 X300.00 Y300.00 Z300.00 E300.00
20:04:55.463 : echo:Maximum Acceleration (mm/s2):
20:04:55.467 : echo:  M201 X3000 Y3000 Z3000 E9000
20:04:55.467 : echo:Accelerations: P=printing, R=retract and T=travel
20:04:55.471 : echo:  M204 P3000.00 R9000.00 T3000.00
20:04:55.475 : echo:Advanced variables: S=Min feedrate (mm/s), T=Min travel feedrate (mm/s), B=minimum segment time (ms), X=maximum XY jerk (mm/s),  Z=maximum Z jerk (mm/s),  E=maximum E jerk (mm/s)
20:04:55.479 : echo:  M205 S0.00 T0.00 B20000 X10.00 Z10.00 E5.00
20:04:55.479 : echo:Home offset (mm)
20:04:55.483 : echo:  M206 X0.00 Y0.00 Z0.00
20:04:55.483 : echo:Endstop adjustment (mm):
20:04:55.483 : echo:  M666 X0.00 Y0.00 Z0.00
20:04:55.487 : echo:Delta settings: L=diagonal_rod, R=radius, S=segments_per_second, ABC=diagonal_rod_trim_tower_[123]
20:04:55.491 : echo:  M665 L217.30 R95.00 S80.00 A0.00 B0.00 C0.00
20:04:55.491 : echo:Material heatup parameters:
20:04:55.491 : echo:  M145 S0 H200 B70 F255
20:04:55.491 : echo:  M145 S1 H240 B100 F255
20:04:55.495 : echo:PID settings:
20:04:55.495 : echo:  M301 P46.03 I6.24 D84.84 C100.00 L20
20:04:55.499 : echo:Retract: S=Length (mm) F:Speed (mm/m) Z: ZLift (mm)
20:04:55.499 : echo:  M207 S3.00 F2700.00 Z0.00
20:04:55.500 : echo:Recover: S=Extra length (mm) F:Speed (mm/m)
20:04:55.504 : echo:  M208 S0.00 F480.00
20:04:55.508 : echo:Auto-Retract: S=0 to disable, 1 to interpret extrude-only moves as retracts or recoveries
20:04:55.508 : echo:  M209 S0
20:04:55.508 : echo:Filament settings: Disabled
20:04:55.508 : echo:  M200 D1.75
20:04:55.508 : echo:  M200 D0
20:04:55.512 : echo:Z-Probe Offset (mm):
20:04:55.512 : echo:  M851 Z0.75
20:04:55.515 : current_position=(0.00, 0.00, 0.00) : sync_plan_position_delta
20:04:55.628 : N1 M110*34
20:04:55.628 : N2 M115*36
20:04:55.628 : N4 M114*35
20:04:55.677 : N5 M111 S7*99
20:04:55.680 : N6 T0*60
20:04:55.680 : N7 M20*22
20:04:55.680 : N8 M80*19
20:05:01.041 : echo:SD init fail
20:05:01.095 : N11 M220 S100*81
20:05:01.095 : FIRMWARE_NAME:Marlin 1.1.0-RCBugFix (Github) SOURCE_CODE_URL:https://github.com/MarlinFirmware/Marlin PROTOCOL_VERSION:1.0 MACHINE_TYPE:Micromake EXTRUDER_COUNT:1 UUID:cede2a2f-41a2-4748-9b12-c55c62f367ff EMERGENCY_CODES:M108,M112,M410
20:05:01.104 : N12 M221 S100*83
20:05:01.104 : X:0.00 Y:0.00 Z:0.00 E:0.00 Count X: 39087 Y:39087 Z:39087
20:05:01.104 : N13 M111 S7*84
20:05:01.104 : echo:DEBUG:ECHO,INFO,ERRORS
20:05:01.104 : echo:N6 T0*60
20:05:01.104 : echo:Active Extruder: 0
20:05:01.104 : echo:N7 M20*22
20:05:01.105 : N14 T0*15
20:05:01.107 : Begin file list
20:05:01.107 : End file list
20:05:01.107 : echo:N8 M80*19
20:05:01.112 : echo:N11 M220 S100*81
20:05:01.115 : echo:N12 M221 S100*83
20:05:01.115 : echo:N13 M111 S7*84
20:05:01.115 : echo:DEBUG:ECHO,INFO,ERRORS
20:05:01.115 : echo:N14 T0*15
20:05:01.119 : echo:Active Extruder: 0
20:05:02.920 : N16 M502*19
20:05:02.925 : echo:N16 M502*19
20:05:02.925 : echo:Hardcoded Default Settings Loaded
20:05:06.211 : N18 M500*31
20:05:06.214 : echo:N18 M500*31
20:05:07.329 : echo:Settings Stored (427 bytes)
20:07:09.971 : N59 G28*47
20:07:09.976 : echo:N59 G28*47
20:07:10.493 : Error:checksum mismatch, Last Line: 59
20:07:10.496 : Resend: 60
20:07:14.002 : X:0.00 Y:0.00 Z:317.06 E:0.00 Count X: 103395 Y:111782 Z:112948
20:07:14.002 : echo:endstops hit:  X:537.91 Y:579.85 Z:585.68
20:07:33.284 : N68 G28*45
20:07:33.289 : echo:N68 G28*45
20:07:45.209 : X:0.00 Y:0.00 Z:317.06 E:0.00 Count X: 147996 Y:148976 Z:134827
20:07:45.209 : echo:endstops hit:  Y:765.82 Z:695.08
```

</details>

| Video clip: first G28 is right, second G28 is wrong | LCD at after first G28, coordinates is right | Repetier-Host at after first G28, coordinates is wrong |
| :-: | :-: | :-: |
| [![](http://img.youtube.com/vi/o29hL11gRNg/0.jpg)](https://www.youtube.com/watch?v=o29hL11gRNg) | ![img_20160719_200727](https://cloud.githubusercontent.com/assets/15308844/16949205/5d717bea-4df4-11e6-8555-cca5a18a033a.png) | ![img_20160719_200715 1](https://cloud.githubusercontent.com/assets/15308844/16949214/660bbe96-4df4-11e6-8303-47de77406911.png) |
|  | LCD at after second G28, coordinates looks right | Repetier-Host at after second G28, coordinates is wrong |
|  | ![img_20160719_200810](https://cloud.githubusercontent.com/assets/15308844/16949222/787a1492-4df4-11e6-81cb-7ed89e19bf27.png) | ![img_20160719_200802 1](https://cloud.githubusercontent.com/assets/15308844/16949224/7b6ceeea-4df4-11e6-8162-db4580f7ef56.png) |

Case 2:
<details>
<summary>

Newest RCBugFix, add PR #4342 (Introduce line_to_axis_pos() for use in homeaxis()), `#define ENDSTOPS_ALWAYS_ON_DEFAULT` is disabled</summary>



```
19:33:47.690 : Printer reset detected - initalizing
19:33:47.691 : start
19:33:47.691 : echo: External Reset
19:33:47.691 : Marlin 1.1.0-RCBugFix
19:33:47.695 : echo: Last Updated: 2016-04-27 12:00 | Author: (Micromake)
19:33:47.695 : Compiled: Jul 19 2016
19:33:47.695 : echo: Free Memory: 2605  PlannerBufferBytes: 1408
19:33:47.699 : echo:V24 stored settings retrieved (427 bytes)
19:33:47.699 : echo:Steps per unit:
19:33:47.703 : echo:  M92 X200.00 Y200.00 Z200.00 E476.55
19:33:47.703 : echo:Maximum feedrates (mm/s):
19:33:47.703 : echo:  M203 X300.00 Y300.00 Z300.00 E300.00
19:33:47.707 : echo:Maximum Acceleration (mm/s2):
19:33:47.707 : echo:  M201 X3000 Y3000 Z3000 E9000
19:33:47.711 : echo:Accelerations: P=printing, R=retract and T=travel
19:33:47.711 : echo:  M204 P3000.00 R9000.00 T3000.00
19:33:47.719 : echo:Advanced variables: S=Min feedrate (mm/s), T=Min travel feedrate (mm/s), B=minimum segment time (ms), X=maximum XY jerk (mm/s),  Z=maximum Z jerk (mm/s),  E=maximum E jerk (mm/s)
19:33:47.720 : echo:  M205 S0.00 T0.00 B20000 X10.00 Z10.00 E5.00
19:33:47.720 : echo:Home offset (mm)
19:33:47.723 : echo:  M206 X0.00 Y0.00 Z0.00
19:33:47.723 : echo:Endstop adjustment (mm):
19:33:47.723 : echo:  M666 X0.00 Y0.00 Z0.00
19:33:47.727 : echo:Delta settings: L=diagonal_rod, R=radius, S=segments_per_second, ABC=diagonal_rod_trim_tower_[123]
19:33:47.731 : echo:  M665 L217.30 R95.00 S80.00 A0.00 B0.00 C0.00
19:33:47.731 : echo:Material heatup parameters:
19:33:47.731 : echo:  M145 S0 H200 B70 F255
19:33:47.735 : echo:  M145 S1 H240 B100 F255
19:33:47.735 : echo:PID settings:
19:33:47.735 : echo:  M301 P46.03 I6.24 D84.84 C100.00 L20
19:33:47.739 : echo:Retract: S=Length (mm) F:Speed (mm/m) Z: ZLift (mm)
19:33:47.740 : echo:  M207 S3.00 F2700.00 Z0.00
19:33:47.743 : echo:Recover: S=Extra length (mm) F:Speed (mm/m)
19:33:47.743 : echo:  M208 S0.00 F480.00
19:33:47.748 : echo:Auto-Retract: S=0 to disable, 1 to interpret extrude-only moves as retracts or recoveries
19:33:47.748 : echo:  M209 S0
19:33:47.748 : echo:Filament settings: Disabled
19:33:47.751 : echo:  M200 D1.75
19:33:47.751 : echo:  M200 D0
19:33:47.752 : echo:Z-Probe Offset (mm):
19:33:47.752 : echo:  M851 Z0.75
19:33:47.755 : current_position=(0.00, 0.00, 0.00) : sync_plan_position_delta
19:33:47.814 : N1 M110*34
19:33:47.814 : N2 M115*36
19:33:47.814 : N4 M114*35
19:33:47.867 : N5 M111 S7*99
19:33:47.868 : N6 T0*60
19:33:47.868 : N7 M20*22
19:33:47.869 : N8 M80*19
19:33:53.284 : echo:SD init fail
19:33:53.284 : N11 M502*20
19:33:53.335 : FIRMWARE_NAME:Marlin 1.1.0-RCBugFix (Github) SOURCE_CODE_URL:https://github.com/MarlinFirmware/Marlin PROTOCOL_VERSION:1.0 MACHINE_TYPE:Micromake EXTRUDER_COUNT:1 UUID:cede2a2f-41a2-4748-9b12-c55c62f367ff EMERGENCY_CODES:M108,M112,M410
19:33:53.345 : N12 M220 S100*82
19:33:53.345 : X:0.00 Y:0.00 Z:0.00 E:0.00 Count X: 39087 Y:39087 Z:39087
19:33:53.345 : N13 M221 S100*82
19:33:53.345 : echo:DEBUG:ECHO,INFO,ERRORS
19:33:53.346 : echo:N6 T0*60
19:33:53.346 : echo:Active Extruder: 0
19:33:53.347 : N14 M111 S7*83
19:33:53.347 : echo:N7 M20*22
19:33:53.347 : Begin file list
19:33:53.348 : N15 T0*14
19:33:53.348 : End file list
19:33:53.348 : echo:N8 M80*19
19:33:53.355 : echo:N11 M502*20
19:33:53.355 : echo:Hardcoded Default Settings Loaded
19:33:53.355 : echo:N12 M220 S100*82
19:33:53.355 : echo:N13 M221 S100*82
19:33:53.359 : echo:N14 M111 S7*83
19:33:53.359 : echo:DEBUG:ECHO,INFO,ERRORS
19:33:53.363 : echo:N15 T0*14
19:33:53.363 : echo:Active Extruder: 0
19:33:54.796 : N17 M500*16
19:33:54.799 : echo:N17 M500*16
19:33:55.915 : echo:Settings Stored (427 bytes)
19:34:19.976 : N26 G28*39
19:34:19.981 : echo:N26 G28*39
19:34:23.533 : Printer reset detected - initalizing
19:34:23.533 : start
19:34:23.533 : echo: External Reset
19:34:23.537 : Marlin 1.1.0-RCBugFix
19:34:23.537 : echo: Last Updated: 2016-04-27 12:00 | Author: (Micromake)
19:34:23.537 : Compiled: Jul 19 2016
19:34:23.541 : echo: Free Memory: 2605  PlannerBufferBytes: 1408
19:34:23.541 : echo:V24 stored settings retrieved (427 bytes)
19:34:23.545 : echo:Steps per unit:
19:34:23.545 : echo:  M92 X200.00 Y200.00 Z200.00 E476.55
19:34:23.545 : echo:Maximum feedrates (mm/s):
19:34:23.549 : echo:  M203 X300.00 Y300.00 Z300.00 E300.00
19:34:23.549 : echo:Maximum Acceleration (mm/s2):
19:34:23.549 : echo:  M201 X3000 Y3000 Z3000 E9000
19:34:23.553 : echo:Accelerations: P=printing, R=retract and T=travel
19:34:23.553 : echo:  M204 P3000.00 R9000.00 T3000.00
19:34:23.562 : echo:Advanced variables: S=Min feedrate (mm/s), T=Min travel feedrate (mm/s), B=minimum segment time (ms), X=maximum XY jerk (mm/s),  Z=maximum Z jerk (mm/s),  E=maximum E jerk (mm/s)
19:34:23.562 : echo:  M205 S0.00 T0.00 B20000 X10.00 Z10.00 E5.00
19:34:23.566 : echo:Home offset (mm)
19:34:23.566 : echo:  M206 X0.00 Y0.00 Z0.00
19:34:23.566 : echo:Endstop adjustment (mm):
19:34:23.570 : echo:  M666 X0.00 Y0.00 Z0.00
19:34:23.574 : echo:Delta settings: L=diagonal_rod, R=radius, S=segments_per_second, ABC=diagonal_rod_trim_tower_[123]
19:34:23.574 : echo:  M665 L217.30 R95.00 S80.00 A0.00 B0.00 C0.00
19:34:23.578 : echo:Material heatup parameters:
19:34:23.578 : echo:  M145 S0 H200 B70 F255
19:34:23.578 : echo:  M145 S1 H240 B100 F255
19:34:23.578 : echo:PID settings:
19:34:23.582 : echo:  M301 P46.03 I6.24 D84.84 C100.00 L20
19:34:23.582 : echo:Retract: S=Length (mm) F:Speed (mm/m) Z: ZLift (mm)
19:34:23.586 : echo:  M207 S3.00 F2700.00 Z0.00
19:34:23.586 : echo:Recover: S=Extra length (mm) F:Speed (mm/m)
19:34:23.586 : echo:  M208 S0.00 F480.00
19:34:23.590 : echo:Auto-Retract: S=0 to disable, 1 to interpret extrude-only moves as retracts or recoveries
19:34:23.590 : echo:  M209 S0
19:34:23.594 : echo:Filament settings: Disabled
19:34:23.594 : echo:  M200 D1.75
19:34:23.594 : echo:  M200 D0
19:34:23.594 : echo:Z-Probe Offset (mm):
19:34:23.595 : echo:  M851 Z0.75
19:34:23.598 : current_position=(0.00, 0.00, 0.00) : sync_plan_position_delta
19:34:23.708 : N1 M110*34
19:34:23.708 : N2 M115*36
19:34:23.708 : N4 M114*35
19:34:23.749 : N5 M111 S7*99
19:34:23.755 : N6 T0*60
19:34:23.755 : N7 M20*22
19:34:23.755 : N8 M80*19
19:34:29.127 : echo:SD init fail
19:34:29.181 : FIRMWARE_NAME:Marlin 1.1.0-RCBugFix (Github) SOURCE_CODE_URL:https://github.com/MarlinFirmware/Marlin PROTOCOL_VERSION:1.0 MACHINE_TYPE:Micromake EXTRUDER_COUNT:1 UUID:cede2a2f-41a2-4748-9b12-c55c62f367ff EMERGENCY_CODES:M108,M112,M410
19:34:29.181 : N11 M220 S100*81
19:34:29.187 : N12 M221 S100*83
19:34:29.187 : X:0.00 Y:0.00 Z:0.00 E:0.00 Count X: 39087 Y:39087 Z:39087
19:34:29.187 : N13 M111 S7*84
19:34:29.187 : echo:DEBUG:ECHO,INFO,ERRORS
19:34:29.187 : echo:N6 T0*60
19:34:29.189 : echo:Active Extruder: 0
19:34:29.189 : echo:N7 M20*22
19:34:29.189 : Begin file list
19:34:29.189 : N14 T0*15
19:34:29.189 : End file list
19:34:29.189 : echo:N8 M80*19
19:34:29.197 : echo:N11 M220 S100*81
19:34:29.198 : Error:checksum mismatch, Last Line: 12
19:34:29.198 : Resend: 13
19:34:29.204 : Resend: N13 M111 S7*84
19:34:29.204 : Resend: N14 T0*15
19:34:29.204 : echo:N12 M221 S100*83
19:34:29.209 : echo:N13 M111 S7*84
19:34:29.209 : echo:DEBUG:ECHO,INFO,ERRORS
19:34:29.210 : echo:N14 T0*15
19:34:29.210 : echo:Active Extruder: 0
```

</details>

Video clip: first G28 is wrong (posted at yesterday)
[![](http://img.youtube.com/vi/VcoUde3wNsI/0.jpg)](https://www.youtube.com/watch?v=VcoUde3wNsI)
A branch that it's used for Case1 and Case2:
https://github.com/esenapaj/Marlin/tree/testes

Case 3 and Case 4 (same result):<details>
<summary>

Newest RCBugFix, `#define ENDSTOPS_ALWAYS_ON_DEFAULT` is enabled</summary>



```
19:24:12.317 : Printer reset detected - initalizing
19:24:12.319 : start
19:24:12.319 : echo: External Reset
19:24:12.319 : Marlin 1.1.0-RCBugFix
19:24:12.323 : echo: Last Updated: 2016-04-27 12:00 | Author: (Micromake)
19:24:12.323 : Compiled: Jul 19 2016
19:24:12.327 : echo: Free Memory: 2604  PlannerBufferBytes: 1408
19:24:12.327 : echo:V24 stored settings retrieved (427 bytes)
19:24:12.327 : echo:Steps per unit:
19:24:12.331 : echo:  M92 X200.00 Y200.00 Z200.00 E476.55
19:24:12.331 : echo:Maximum feedrates (mm/s):
19:24:12.331 : echo:  M203 X300.00 Y300.00 Z300.00 E300.00
19:24:12.335 : echo:Maximum Acceleration (mm/s2):
19:24:12.335 : echo:  M201 X3000 Y3000 Z3000 E9000
19:24:12.339 : echo:Accelerations: P=printing, R=retract and T=travel
19:24:12.339 : echo:  M204 P3000.00 R9000.00 T3000.00
19:24:12.347 : echo:Advanced variables: S=Min feedrate (mm/s), T=Min travel feedrate (mm/s), B=minimum segment time (ms), X=maximum XY jerk (mm/s),  Z=maximum Z jerk (mm/s),  E=maximum E jerk (mm/s)
19:24:12.347 : echo:  M205 S0.00 T0.00 B20000 X10.00 Z10.00 E5.00
19:24:12.351 : echo:Home offset (mm)
19:24:12.351 : echo:  M206 X0.00 Y0.00 Z0.00
19:24:12.351 : echo:Endstop adjustment (mm):
19:24:12.355 : echo:  M666 X0.00 Y0.00 Z0.00
19:24:12.355 : echo:Delta settings: L=diagonal_rod, R=radius, S=segments_per_second, ABC=diagonal_rod_trim_tower_[123]
19:24:12.359 : echo:  M665 L217.30 R95.00 S80.00 A0.00 B0.00 C0.00
19:24:12.359 : echo:Material heatup parameters:
19:24:12.363 : echo:  M145 S0 H200 B70 F255
19:24:12.363 : echo:  M145 S1 H240 B100 F255
19:24:12.363 : echo:PID settings:
19:24:12.367 : echo:  M301 P46.03 I6.24 D84.84 C100.00 L20
19:24:12.367 : echo:Retract: S=Length (mm) F:Speed (mm/m) Z: ZLift (mm)
19:24:12.371 : echo:  M207 S3.00 F2700.00 Z0.00
19:24:12.371 : echo:Recover: S=Extra length (mm) F:Speed (mm/m)
19:24:12.371 : echo:  M208 S0.00 F480.00
19:24:12.375 : echo:Auto-Retract: S=0 to disable, 1 to interpret extrude-only moves as retracts or recoveries
19:24:12.375 : echo:  M209 S0
19:24:12.379 : echo:Filament settings: Disabled
19:24:12.379 : echo:  M200 D1.75
19:24:12.379 : echo:  M200 D0
19:24:12.379 : echo:Z-Probe Offset (mm):
19:24:12.379 : echo:  M851 Z0.75
19:24:12.383 : current_position=(0.00, 0.00, 0.00) : sync_plan_position_delta
19:24:12.518 : N1 M110*34
19:24:12.518 : N2 M115*36
19:24:12.518 : N4 M114*35
19:24:12.577 : N5 M111 S7*99
19:24:12.580 : N6 T0*60
19:24:12.580 : N7 M20*22
19:24:12.580 : N8 M80*19
19:24:17.912 : echo:SD init fail
19:24:17.967 : N11 M220 S100*81
19:24:17.966 : FIRMWARE_NAME:Marlin 1.1.0-RCBugFix (Github) SOURCE_CODE_URL:https://github.com/MarlinFirmware/Marlin PROTOCOL_VERSION:1.0 MACHINE_TYPE:Micromake EXTRUDER_COUNT:1 UUID:cede2a2f-41a2-4748-9b12-c55c62f367ff EMERGENCY_CODES:M108,M112,M410
19:24:17.974 : N12 M221 S100*83
19:24:17.974 : X:0.00 Y:0.00 Z:0.00 E:0.00 Count X: 39087 Y:39087 Z:39087
19:24:17.974 : N13 M111 S7*84
19:24:17.975 : echo:DEBUG:ECHO,INFO,ERRORS
19:24:17.975 : echo:N6 T0*60
19:24:17.975 : echo:Active Extruder: 0
19:24:17.975 : echo:N7 M20*22
19:24:17.975 : Begin file list
19:24:17.976 : End file list
19:24:17.976 : echo:N8 M80*19
19:24:17.976 : N14 T0*15
19:24:17.984 : echo:N11 M220 S100*81
19:24:17.984 : echo:N12 M221 S100*83
19:24:17.984 : echo:N13 M111 S7*84
19:24:17.986 : echo:DEBUG:ECHO,INFO,ERRORS
19:24:17.986 : echo:N14 T0*15
19:24:17.986 : echo:Active Extruder: 0
19:24:28.536 : N19 G28*43
19:24:28.541 : echo:N19 G28*43
19:24:31.159 : X:0.00 Y:0.00 Z:317.06 E:0.00 Count X: 102498 Y:102498 Z:102498
19:24:35.188 : N22 G28*35
19:24:35.193 : echo:N22 G28*35
19:24:36.229 : X:0.00 Y:0.00 Z:317.06 E:0.00 Count X: 102498 Y:102498 Z:102498
```

</details>

<details>
<summary>

Newest RCBugFix, `#define ENDSTOPS_ALWAYS_ON_DEFAULT` is disabled</summary>



```
19:18:33.527 : Printer reset detected - initalizing
19:18:33.527 : start
19:18:33.532 : echo: External Reset
19:18:33.532 : Marlin 1.1.0-RCBugFix
19:18:33.532 : echo: Last Updated: 2016-04-27 12:00 | Author: (Micromake)
19:18:33.532 : Compiled: Jul 19 2016
19:18:33.536 : echo: Free Memory: 2605  PlannerBufferBytes: 1408
19:18:33.540 : echo:V24 stored settings retrieved (427 bytes)
19:18:33.540 : echo:Steps per unit:
19:18:33.540 : echo:  M92 X200.00 Y200.00 Z200.00 E476.55
19:18:33.541 : echo:Maximum feedrates (mm/s):
19:18:33.544 : echo:  M203 X300.00 Y300.00 Z300.00 E300.00
19:18:33.544 : echo:Maximum Acceleration (mm/s2):
19:18:33.548 : echo:  M201 X3000 Y3000 Z3000 E9000
19:18:33.548 : echo:Accelerations: P=printing, R=retract and T=travel
19:18:33.549 : echo:  M204 P3000.00 R9000.00 T3000.00
19:18:33.557 : echo:Advanced variables: S=Min feedrate (mm/s), T=Min travel feedrate (mm/s), B=minimum segment time (ms), X=maximum XY jerk (mm/s),  Z=maximum Z jerk (mm/s),  E=maximum E jerk (mm/s)
19:18:33.561 : echo:  M205 S0.00 T0.00 B20000 X10.00 Z10.00 E5.00
19:18:33.561 : echo:Home offset (mm)
19:18:33.561 : echo:  M206 X0.00 Y0.00 Z0.00
19:18:33.564 : echo:Endstop adjustment (mm):
19:18:33.565 : echo:  M666 X0.00 Y0.00 Z0.00
19:18:33.569 : echo:Delta settings: L=diagonal_rod, R=radius, S=segments_per_second, ABC=diagonal_rod_trim_tower_[123]
19:18:33.569 : echo:  M665 L217.30 R95.00 S80.00 A0.00 B0.00 C0.00
19:18:33.573 : echo:Material heatup parameters:
19:18:33.573 : echo:  M145 S0 H200 B70 F255
19:18:33.573 : echo:  M145 S1 H240 B100 F255
19:18:33.573 : echo:PID settings:
19:18:33.577 : echo:  M301 P46.03 I6.24 D84.84 C100.00 L20
19:18:33.581 : echo:Retract: S=Length (mm) F:Speed (mm/m) Z: ZLift (mm)
19:18:33.581 : echo:  M207 S3.00 F2700.00 Z0.00
19:18:33.581 : echo:Recover: S=Extra length (mm) F:Speed (mm/m)
19:18:33.581 : echo:  M208 S0.00 F480.00
19:18:33.589 : echo:Auto-Retract: S=0 to disable, 1 to interpret extrude-only moves as retracts or recoveries
19:18:33.589 : echo:  M209 S0
19:18:33.589 : echo:Filament settings: Disabled
19:18:33.589 : echo:  M200 D1.75
19:18:33.590 : echo:  M200 D0
19:18:33.590 : echo:Z-Probe Offset (mm):
19:18:33.593 : echo:  M851 Z0.75
19:18:33.593 : current_position=(0.00, 0.00, 0.00) : sync_plan_position_delta
19:18:33.674 : N1 M110*34
19:18:33.674 : N2 M115*36
19:18:33.674 : N4 M114*35
19:18:33.722 : N5 M111 S7*99
19:18:33.725 : N6 T0*60
19:18:33.725 : N7 M20*22
19:18:33.725 : N8 M80*19
19:18:39.122 : echo:SD init fail
19:18:39.176 : FIRMWARE_NAME:Marlin 1.1.0-RCBugFix (Github) SOURCE_CODE_URL:https://github.com/MarlinFirmware/Marlin PROTOCOL_VERSION:1.0 MACHINE_TYPE:Micromake EXTRUDER_COUNT:1 UUID:cede2a2f-41a2-4748-9b12-c55c62f367ff EMERGENCY_CODES:M108,M112,M410
19:18:39.176 : N11 M220 S100*81
19:18:39.180 : N12 M221 S100*83
19:18:39.180 : X:0.00 Y:0.00 Z:0.00 E:0.00 Count X: 39087 Y:39087 Z:39087
19:18:39.180 : N13 M111 S7*84
19:18:39.180 : echo:DEBUG:ECHO,INFO,ERRORS
19:18:39.181 : N14 T0*15
19:18:39.184 : echo:N6 T0*60
19:18:39.184 : echo:Active Extruder: 0
19:18:39.184 : echo:N7 M20*22
19:18:39.184 : Begin file list
19:18:39.185 : End file list
19:18:39.188 : echo:N8 M80*19
19:18:39.192 : echo:N11 M220 S100*81
19:18:39.192 : echo:N12 M221 S100*83
19:18:39.196 : echo:N13 M111 S7*84
19:18:39.196 : echo:DEBUG:ECHO,INFO,ERRORS
19:18:39.196 : echo:N14 T0*15
19:18:39.196 : echo:Active Extruder: 0
19:18:52.661 : N20 G28*33
19:18:52.666 : echo:N20 G28*33
19:18:55.440 : X:0.00 Y:0.00 Z:317.06 E:0.00 Count X: 102498 Y:102498 Z:102498
19:18:57.615 : N22 G28*35
19:18:57.618 : echo:N22 G28*35
19:18:58.631 : X:0.00 Y:0.00 Z:317.06 E:0.00 Count X: 102498 Y:102498 Z:102498
```

</details>

| Video clip: first G28 is wrong, second G28 is also wrong | LCD at after first and second G28 (same result), coordinates looks right | Repetier-Host at after first and second G28 (same result), coordinates looks right |
| :-: | :-: | :-: |
| [![](http://img.youtube.com/vi/yPseYcALuqg/0.jpg)](https://www.youtube.com/watch?v=yPseYcALuqg) | ![dscn1694](https://cloud.githubusercontent.com/assets/15308844/16949452/deab9dd4-4df5-11e6-995c-b40cfeffa796.png) | ![dscn1695](https://cloud.githubusercontent.com/assets/15308844/16949459/e3101f30-4df5-11e6-915f-bca0e027a39b.png) |

A branch that it's used for Case3 and Case4:
https://github.com/esenapaj/Marlin/tree/testes2

ghost commented:
Case5 and Case6 (same result):

<details>
<summary>

Newest RCBugFix, revert PR #4338 (Use blocking_move, current_position in G28 homing), `#define ENDSTOPS_ALWAYS_ON_DEFAULT` is enabled</summary>



```
21:43:44.031 : Printer reset detected - initalizing
21:43:44.031 : start
21:43:44.031 : echo: External Reset
21:43:44.031 : Marlin 1.1.0-RCBugFix
21:43:44.035 : echo: Last Updated: 2016-04-27 12:00 | Author: (Micromake)
21:43:44.036 : Compiled: Jul 19 2016
21:43:44.040 : echo: Free Memory: 2604  PlannerBufferBytes: 1408
21:43:44.040 : echo:V24 stored settings retrieved (427 bytes)
21:43:44.040 : echo:Steps per unit:
21:43:44.044 : echo:  M92 X200.00 Y200.00 Z200.00 E476.55
21:43:44.044 : echo:Maximum feedrates (mm/s):
21:43:44.044 : echo:  M203 X300.00 Y300.00 Z300.00 E300.00
21:43:44.048 : echo:Maximum Acceleration (mm/s2):
21:43:44.048 : echo:  M201 X3000 Y3000 Z3000 E9000
21:43:44.052 : echo:Accelerations: P=printing, R=retract and T=travel
21:43:44.052 : echo:  M204 P3000.00 R9000.00 T3000.00
21:43:44.060 : echo:Advanced variables: S=Min feedrate (mm/s), T=Min travel feedrate (mm/s), B=minimum segment time (ms), X=maximum XY jerk (mm/s),  Z=maximum Z jerk (mm/s),  E=maximum E jerk (mm/s)
21:43:44.060 : echo:  M205 S0.00 T0.00 B20000 X10.00 Z10.00 E5.00
21:43:44.060 : echo:Home offset (mm)
21:43:44.064 : echo:  M206 X0.00 Y0.00 Z0.00
21:43:44.064 : echo:Endstop adjustment (mm):
21:43:44.068 : echo:  M666 X0.00 Y0.00 Z0.00
21:43:44.068 : echo:Delta settings: L=diagonal_rod, R=radius, S=segments_per_second, ABC=diagonal_rod_trim_tower_[123]
21:43:44.072 : echo:  M665 L217.30 R95.00 S80.00 A0.00 B0.00 C0.00
21:43:44.073 : echo:Material heatup parameters:
21:43:44.076 : echo:  M145 S0 H200 B70 F255
21:43:44.076 : echo:  M145 S1 H240 B100 F255
21:43:44.076 : echo:PID settings:
21:43:44.077 : echo:  M301 P46.03 I6.24 D84.84 C100.00 L20
21:43:44.081 : echo:Retract: S=Length (mm) F:Speed (mm/m) Z: ZLift (mm)
21:43:44.081 : echo:  M207 S3.00 F2700.00 Z0.00
21:43:44.085 : echo:Recover: S=Extra length (mm) F:Speed (mm/m)
21:43:44.085 : echo:  M208 S0.00 F480.00
21:43:44.089 : echo:Auto-Retract: S=0 to disable, 1 to interpret extrude-only moves as retracts or recoveries
21:43:44.089 : echo:  M209 S0
21:43:44.093 : echo:Filament settings: Disabled
21:43:44.093 : echo:  M200 D1.75
21:43:44.093 : echo:  M200 D0
21:43:44.093 : echo:Z-Probe Offset (mm):
21:43:44.093 : echo:  M851 Z0.75
21:43:44.097 : current_position=(0.00, 0.00, 0.00) : sync_plan_position_delta
21:43:44.240 : N1 M110*34
21:43:44.240 : N2 M115*36
21:43:44.240 : N4 M114*35
21:43:44.263 : N5 M111 S6*98
21:43:44.265 : N6 T0*60
21:43:44.265 : N7 M20*22
21:43:44.265 : N8 M80*19
21:43:49.626 : echo:SD init fail
21:43:49.626 : N11 M502*20
21:43:49.680 : FIRMWARE_NAME:Marlin 1.1.0-RCBugFix (Github) SOURCE_CODE_URL:https://github.com/MarlinFirmware/Marlin PROTOCOL_VERSION:1.0 MACHINE_TYPE:Micromake EXTRUDER_COUNT:1 UUID:cede2a2f-41a2-4748-9b12-c55c62f367ff EMERGENCY_CODES:M108,M112,M410
21:43:49.685 : N12 M220 S100*82
21:43:49.685 : X:0.00 Y:0.00 Z:0.00 E:0.00 Count X: 39087 Y:39087 Z:39087
21:43:49.685 : N13 M221 S100*82
21:43:49.685 : echo:DEBUG:INFO,ERRORS
21:43:49.685 : echo:Active Extruder: 0
21:43:49.686 : N14 M111 S6*82
21:43:49.686 : N15 T0*14
21:43:49.688 : Begin file list
21:43:49.689 : End file list
21:43:49.693 : echo:Hardcoded Default Settings Loaded
21:43:49.696 : echo:DEBUG:INFO,ERRORS
21:43:49.696 : echo:Active Extruder: 0
21:43:52.387 : N17 M500*16
21:43:53.506 : echo:Settings Stored (427 bytes)
21:43:59.935 : N21 G28*32
21:44:02.708 : X:0.00 Y:0.00 Z:317.06 E:0.00 Count X: 102498 Y:102498 Z:102498
21:44:09.656 : N25 G28*36
21:44:12.445 : X:0.00 Y:0.00 Z:317.06 E:0.00 Count X: 102498 Y:102498 Z:102498
```

</details>

<details>
<summary>

Newest RCBugFix, revert PR #4338 (Use blocking_move, current_position in G28 homing), `#define ENDSTOPS_ALWAYS_ON_DEFAULT` is disabled</summary>



```
21:39:04.425 : Printer reset detected - initalizing
21:39:04.425 : start
21:39:04.429 : echo: External Reset
21:39:04.430 : Marlin 1.1.0-RCBugFix
21:39:04.430 : echo: Last Updated: 2016-04-27 12:00 | Author: (Micromake)
21:39:04.434 : Compiled: Jul 19 2016
21:39:04.434 : echo: Free Memory: 2605  PlannerBufferBytes: 1408
21:39:04.438 : echo:V24 stored settings retrieved (427 bytes)
21:39:04.438 : echo:Steps per unit:
21:39:04.438 : echo:  M92 X200.00 Y200.00 Z200.00 E476.55
21:39:04.438 : echo:Maximum feedrates (mm/s):
21:39:04.442 : echo:  M203 X300.00 Y300.00 Z300.00 E300.00
21:39:04.442 : echo:Maximum Acceleration (mm/s2):
21:39:04.446 : echo:  M201 X3000 Y3000 Z3000 E9000
21:39:04.446 : echo:Accelerations: P=printing, R=retract and T=travel
21:39:04.446 : echo:  M204 P3000.00 R9000.00 T3000.00
21:39:04.454 : echo:Advanced variables: S=Min feedrate (mm/s), T=Min travel feedrate (mm/s), B=minimum segment time (ms), X=maximum XY jerk (mm/s),  Z=maximum Z jerk (mm/s),  E=maximum E jerk (mm/s)
21:39:04.458 : echo:  M205 S0.00 T0.00 B20000 X10.00 Z10.00 E5.00
21:39:04.458 : echo:Home offset (mm)
21:39:04.458 : echo:  M206 X0.00 Y0.00 Z0.00
21:39:04.462 : echo:Endstop adjustment (mm):
21:39:04.462 : echo:  M666 X0.00 Y0.00 Z0.00
21:39:04.466 : echo:Delta settings: L=diagonal_rod, R=radius, S=segments_per_second, ABC=diagonal_rod_trim_tower_[123]
21:39:04.470 : echo:  M665 L217.30 R95.00 S80.00 A0.00 B0.00 C0.00
21:39:04.470 : echo:Material heatup parameters:
21:39:04.470 : echo:  M145 S0 H200 B70 F255
21:39:04.470 : echo:  M145 S1 H240 B100 F255
21:39:04.471 : echo:PID settings:
21:39:04.474 : echo:  M301 P46.03 I6.24 D84.84 C100.00 L20
21:39:04.478 : echo:Retract: S=Length (mm) F:Speed (mm/m) Z: ZLift (mm)
21:39:04.479 : echo:  M207 S3.00 F2700.00 Z0.00
21:39:04.479 : echo:Recover: S=Extra length (mm) F:Speed (mm/m)
21:39:04.483 : echo:  M208 S0.00 F480.00
21:39:04.487 : echo:Auto-Retract: S=0 to disable, 1 to interpret extrude-only moves as retracts or recoveries
21:39:04.487 : echo:  M209 S0
21:39:04.487 : echo:Filament settings: Disabled
21:39:04.487 : echo:  M200 D1.75
21:39:04.487 : echo:  M200 D0
21:39:04.491 : echo:Z-Probe Offset (mm):
21:39:04.491 : echo:  M851 Z0.75
21:39:04.491 : current_position=(0.00, 0.00, 0.00) : sync_plan_position_delta
21:39:04.612 : N1 M110*34
21:39:04.612 : N2 M115*36
21:39:04.612 : N4 M114*35
21:39:04.668 : N5 M111 S6*98
21:39:04.671 : N6 T0*60
21:39:04.671 : N7 M20*22
21:39:04.671 : N8 M80*19
21:39:10.020 : echo:SD init fail
21:39:10.074 : FIRMWARE_NAME:Marlin 1.1.0-RCBugFix (Github) SOURCE_CODE_URL:https://github.com/MarlinFirmware/Marlin PROTOCOL_VERSION:1.0 MACHINE_TYPE:Micromake EXTRUDER_COUNT:1 UUID:cede2a2f-41a2-4748-9b12-c55c62f367ff EMERGENCY_CODES:M108,M112,M410
21:39:10.074 : N11 M220 S100*81
21:39:10.079 : N12 M221 S100*83
21:39:10.079 : X:0.00 Y:0.00 Z:0.00 E:0.00 Count X: 39087 Y:39087 Z:39087
21:39:10.079 : echo:DEBUG:INFO,ERRORS
21:39:10.079 : N13 M111 S6*85
21:39:10.080 : N14 T0*15
21:39:10.082 : echo:Active Extruder: 0
21:39:10.082 : Begin file list
21:39:10.082 : End file list
21:39:10.089 : echo:DEBUG:INFO,ERRORS
21:39:10.090 : echo:Active Extruder: 0
21:39:10.368 : N15 M502*16
21:39:10.372 : echo:Hardcoded Default Settings Loaded
21:39:13.367 : N17 M500*16
21:39:14.488 : echo:Settings Stored (427 bytes)
21:39:16.483 : N19 G28*43
21:39:19.256 : X:0.00 Y:0.00 Z:317.06 E:0.00 Count X: 102498 Y:102498 Z:102498
21:39:26.941 : N24 G28*37
21:39:29.729 : X:0.00 Y:0.00 Z:317.06 E:0.00 Count X: 102498 Y:102498 Z:102498
```

</details>
No problem

A branch that it's used for Case5 and Case6:
https://github.com/esenapaj/Marlin/tree/testes3

ghost commented:
Case 7:
<details>
<summary>

Newest RCBugFix, revert PR #4306 (ENDSTOPS_ONLY_FOR_HOMING is confusing), add PR #4342 (Introduce line_to_axis_pos() for use in homeaxis()), `#define ENDSTOPS_ONLY_FOR_HOMING` is enabled</summary>



```
22:18:11.332 : Printer reset detected - initalizing
22:18:11.332 : start
22:18:11.336 : echo: External Reset
22:18:11.336 : Marlin 1.1.0-RCBugFix
22:18:11.336 : echo: Last Updated: 2016-04-27 12:00 | Author: (Micromake)
22:18:11.336 : Compiled: Jul 19 2016
22:18:11.340 : echo: Free Memory: 2605  PlannerBufferBytes: 1408
22:18:11.340 : echo:V24 stored settings retrieved (427 bytes)
22:18:11.344 : echo:Steps per unit:
22:18:11.344 : echo:  M92 X200.00 Y200.00 Z200.00 E476.55
22:18:11.344 : echo:Maximum feedrates (mm/s):
22:18:11.349 : echo:  M203 X300.00 Y300.00 Z300.00 E300.00
22:18:11.349 : echo:Maximum Acceleration (mm/s2):
22:18:11.352 : echo:  M201 X3000 Y3000 Z3000 E9000
22:18:11.353 : echo:Accelerations: P=printing, R=retract and T=travel
22:18:11.353 : echo:  M204 P3000.00 R9000.00 T3000.00
22:18:11.361 : echo:Advanced variables: S=Min feedrate (mm/s), T=Min travel feedrate (mm/s), B=minimum segment time (ms), X=maximum XY jerk (mm/s),  Z=maximum Z jerk (mm/s),  E=maximum E jerk (mm/s)
22:18:11.365 : echo:  M205 S0.00 T0.00 B20000 X10.00 Z10.00 E5.00
22:18:11.366 : echo:Home offset (mm)
22:18:11.366 : echo:  M206 X0.00 Y0.00 Z0.00
22:18:11.369 : echo:Endstop adjustment (mm):
22:18:11.369 : echo:  M666 X0.00 Y0.00 Z0.00
22:18:11.374 : echo:Delta settings: L=diagonal_rod, R=radius, S=segments_per_second, ABC=diagonal_rod_trim_tower_[123]
22:18:11.374 : echo:  M665 L217.30 R95.00 S80.00 A0.00 B0.00 C0.00
22:18:11.377 : echo:Material heatup parameters:
22:18:11.377 : echo:  M145 S0 H200 B70 F255
22:18:11.377 : echo:  M145 S1 H240 B100 F255
22:18:11.377 : echo:PID settings:
22:18:11.381 : echo:  M301 P46.03 I6.24 D84.84 C100.00 L20
22:18:11.381 : echo:Retract: S=Length (mm) F:Speed (mm/m) Z: ZLift (mm)
22:18:11.385 : echo:  M207 S3.00 F2700.00 Z0.00
22:18:11.385 : echo:Recover: S=Extra length (mm) F:Speed (mm/m)
22:18:11.386 : echo:  M208 S0.00 F480.00
22:18:11.390 : echo:Auto-Retract: S=0 to disable, 1 to interpret extrude-only moves as retracts or recoveries
22:18:11.393 : echo:  M209 S0
22:18:11.393 : echo:Filament settings: Disabled
22:18:11.393 : echo:  M200 D1.75
22:18:11.393 : echo:  M200 D0
22:18:11.394 : echo:Z-Probe Offset (mm):
22:18:11.397 : echo:  M851 Z0.75
22:18:11.398 : current_position=(0.00, 0.00, 0.00) : sync_plan_position_delta
22:18:11.443 : N1 M110*34
22:18:11.443 : N2 M115*36
22:18:11.443 : N4 M114*35
22:18:11.468 : N5 M111 S6*98
22:18:11.472 : N6 T0*60
22:18:11.472 : N7 M20*22
22:18:11.472 : N8 M80*19
22:18:16.927 : echo:SD init fail
22:18:16.981 : N11 M220 S100*81
22:18:16.981 : FIRMWARE_NAME:Marlin 1.1.0-RCBugFix (Github) SOURCE_CODE_URL:https://github.com/MarlinFirmware/Marlin PROTOCOL_VERSION:1.0 MACHINE_TYPE:Micromake EXTRUDER_COUNT:1 UUID:cede2a2f-41a2-4748-9b12-c55c62f367ff EMERGENCY_CODES:M108,M112,M410
22:18:16.990 : N12 M221 S100*83
22:18:16.990 : X:0.00 Y:0.00 Z:0.00 E:0.00 Count X: 39087 Y:39087 Z:39087
22:18:16.990 : N13 M111 S6*85
22:18:16.991 : echo:DEBUG:INFO,ERRORS
22:18:16.991 : echo:Active Extruder: 0
22:18:16.991 : Begin file list
22:18:16.992 : End file list
22:18:16.992 : N14 T0*15
22:18:16.998 : echo:DEBUG:INFO,ERRORS
22:18:16.998 : echo:Unknown command: "2 P15 B35"
22:18:17.566 : Error:Line Number is not Last Line Number+1, Last Line: 13
22:18:17.570 : Resend: 14
22:18:17.587 : Resend: N14 T0*15
22:18:17.591 : echo:Active Extruder: 0
22:18:19.103 : N16 M502*19
22:18:19.110 : echo:Hardcoded Default Settings Loaded
22:18:22.392 : N18 M500*31
22:18:23.513 : echo:Settings Stored (427 bytes)
22:19:04.476 : N33 G28*35
22:19:07.354 : Printer reset detected - initalizing
22:19:07.355 : start
22:19:07.355 : echo: External Reset
22:19:07.359 : Marlin 1.1.0-RCBugFix
22:19:07.359 : echo: Last Updated: 2016-04-27 12:00 | Author: (Micromake)
22:19:07.359 : Compiled: Jul 19 2016
22:19:07.363 : echo: Free Memory: 2605  PlannerBufferBytes: 1408
22:19:07.363 : echo:V24 stored settings retrieved (427 bytes)
22:19:07.367 : echo:Steps per unit:
22:19:07.367 : echo:  M92 X200.00 Y200.00 Z200.00 E476.55
22:19:07.367 : echo:Maximum feedrates (mm/s):
22:19:07.371 : echo:  M203 X300.00 Y300.00 Z300.00 E300.00
22:19:07.371 : echo:Maximum Acceleration (mm/s2):
22:19:07.371 : echo:  M201 X3000 Y3000 Z3000 E9000
22:19:07.375 : echo:Accelerations: P=printing, R=retract and T=travel
22:19:07.375 : echo:  M204 P3000.00 R9000.00 T3000.00
22:19:07.384 : echo:Advanced variables: S=Min feedrate (mm/s), T=Min travel feedrate (mm/s), B=minimum segment time (ms), X=maximum XY jerk (mm/s),  Z=maximum Z jerk (mm/s),  E=maximum E jerk (mm/s)
22:19:07.388 : echo:  M205 S0.00 T0.00 B20000 X10.00 Z10.00 E5.00
22:19:07.388 : echo:Home offset (mm)
22:19:07.388 : echo:  M206 X0.00 Y0.00 Z0.00
22:19:07.391 : echo:Endstop adjustment (mm):
22:19:07.391 : echo:  M666 X0.00 Y0.00 Z0.00
22:19:07.396 : echo:Delta settings: L=diagonal_rod, R=radius, S=segments_per_second, ABC=diagonal_rod_trim_tower_[123]
22:19:07.396 : echo:  M665 L217.30 R95.00 S80.00 A0.00 B0.00 C0.00
22:19:07.400 : echo:Material heatup parameters:
22:19:07.400 : echo:  M145 S0 H200 B70 F255
22:19:07.400 : echo:  M145 S1 H240 B100 F255
22:19:07.400 : echo:PID settings:
22:19:07.404 : echo:  M301 P46.03 I6.24 D84.84 C100.00 L20
22:19:07.404 : echo:Retract: S=Length (mm) F:Speed (mm/m) Z: ZLift (mm)
22:19:07.408 : echo:  M207 S3.00 F2700.00 Z0.00
22:19:07.408 : echo:Recover: S=Extra length (mm) F:Speed (mm/m)
22:19:07.408 : echo:  M208 S0.00 F480.00
22:19:07.412 : echo:Auto-Retract: S=0 to disable, 1 to interpret extrude-only moves as retracts or recoveries
22:19:07.416 : echo:  M209 S0
22:19:07.416 : echo:Filament settings: Disabled
22:19:07.416 : echo:  M200 D1.75
22:19:07.416 : echo:  M200 D0
22:19:07.416 : echo:Z-Probe Offset (mm):
22:19:07.416 : echo:  M851 Z0.75
22:19:07.420 : current_position=(0.00, 0.00, 0.00) : sync_plan_position_delta
22:19:07.564 : N1 M110*34
22:19:07.564 : N2 M115*36
22:19:07.564 : N4 M114*35
22:19:07.605 : N5 M111 S6*98
22:19:07.613 : N6 T0*60
22:19:07.613 : N7 M20*22
22:19:07.614 : N8 M80*19
22:19:12.949 : echo:SD init fail
22:19:13.003 : N11 M220 S100*81
22:19:13.003 : FIRMWARE_NAME:Marlin 1.1.0-RCBugFix (Github) SOURCE_CODE_URL:https://github.com/MarlinFirmware/Marlin PROTOCOL_VERSION:1.0 MACHINE_TYPE:Micromake EXTRUDER_COUNT:1 UUID:cede2a2f-41a2-4748-9b12-c55c62f367ff EMERGENCY_CODES:M108,M112,M410
22:19:13.008 : N12 M221 S100*83
22:19:13.008 : X:0.00 Y:0.00 Z:0.00 E:0.00 Count X: 39087 Y:39087 Z:39087
22:19:13.008 : N13 M111 S6*85
22:19:13.008 : echo:DEBUG:INFO,ERRORS
22:19:13.009 : N14 T0*15
22:19:13.011 : echo:Active Extruder: 0
22:19:13.011 : Begin file list
22:19:13.011 : End file list
22:19:13.016 : echo:DEBUG:INFO,ERRORS
22:19:13.019 : echo:Active Extruder: 0
```

</details>
Result is same as Case2

Case 8:
<details>
<summary>

Newest RCBugFix, revert PR #4306 (ENDSTOPS_ONLY_FOR_HOMING is confusing), add PR #4342 (Introduce line_to_axis_pos() for use in homeaxis()), `#define ENDSTOPS_ONLY_FOR_HOMING` is disabled</summary>



```
22:22:10.370 : Printer reset detected - initalizing
22:22:10.370 : start
22:22:10.370 : echo: External Reset
22:22:10.370 : Marlin 1.1.0-RCBugFix
22:22:10.374 : echo: Last Updated: 2016-04-27 12:00 | Author: (Micromake)
22:22:10.374 : Compiled: Jul 19 2016
22:22:10.375 : echo: Free Memory: 2604  PlannerBufferBytes: 1408
22:22:10.379 : echo:V24 stored settings retrieved (427 bytes)
22:22:10.379 : echo:Steps per unit:
22:22:10.383 : echo:  M92 X200.00 Y200.00 Z200.00 E476.55
22:22:10.383 : echo:Maximum feedrates (mm/s):
22:22:10.383 : echo:  M203 X300.00 Y300.00 Z300.00 E300.00
22:22:10.387 : echo:Maximum Acceleration (mm/s2):
22:22:10.387 : echo:  M201 X3000 Y3000 Z3000 E9000
22:22:10.391 : echo:Accelerations: P=printing, R=retract and T=travel
22:22:10.391 : echo:  M204 P3000.00 R9000.00 T3000.00
22:22:10.400 : echo:Advanced variables: S=Min feedrate (mm/s), T=Min travel feedrate (mm/s), B=minimum segment time (ms), X=maximum XY jerk (mm/s),  Z=maximum Z jerk (mm/s),  E=maximum E jerk (mm/s)
22:22:10.400 : echo:  M205 S0.00 T0.00 B20000 X10.00 Z10.00 E5.00
22:22:10.400 : echo:Home offset (mm)
22:22:10.403 : echo:  M206 X0.00 Y0.00 Z0.00
22:22:10.404 : echo:Endstop adjustment (mm):
22:22:10.404 : echo:  M666 X0.00 Y0.00 Z0.00
22:22:10.408 : echo:Delta settings: L=diagonal_rod, R=radius, S=segments_per_second, ABC=diagonal_rod_trim_tower_[123]
22:22:10.412 : echo:  M665 L217.30 R95.00 S80.00 A0.00 B0.00 C0.00
22:22:10.412 : echo:Material heatup parameters:
22:22:10.412 : echo:  M145 S0 H200 B70 F255
22:22:10.416 : echo:  M145 S1 H240 B100 F255
22:22:10.416 : echo:PID settings:
22:22:10.416 : echo:  M301 P46.03 I6.24 D84.84 C100.00 L20
22:22:10.420 : echo:Retract: S=Length (mm) F:Speed (mm/m) Z: ZLift (mm)
22:22:10.421 : echo:  M207 S3.00 F2700.00 Z0.00
22:22:10.424 : echo:Recover: S=Extra length (mm) F:Speed (mm/m)
22:22:10.424 : echo:  M208 S0.00 F480.00
22:22:10.429 : echo:Auto-Retract: S=0 to disable, 1 to interpret extrude-only moves as retracts or recoveries
22:22:10.429 : echo:  M209 S0
22:22:10.429 : echo:Filament settings: Disabled
22:22:10.432 : echo:  M200 D1.75
22:22:10.433 : echo:  M200 D0
22:22:10.434 : echo:Z-Probe Offset (mm):
22:22:10.434 : echo:  M851 Z0.75
22:22:10.435 : current_position=(0.00, 0.00, 0.00) : sync_plan_position_delta
22:22:10.583 : N1 M110*34
22:22:10.583 : N2 M115*36
22:22:10.583 : N4 M114*35
22:22:10.620 : N5 M111 S6*98
22:22:10.622 : N6 T0*60
22:22:10.622 : N7 M20*22
22:22:10.622 : N8 M80*19
22:22:15.965 : echo:SD init fail
22:22:15.965 : N11 M502*20
22:22:16.015 : FIRMWARE_NAME:Marlin 1.1.0-RCBugFix (Github) SOURCE_CODE_URL:https://github.com/MarlinFirmware/Marlin PROTOCOL_VERSION:1.0 MACHINE_TYPE:Micromake EXTRUDER_COUNT:1 UUID:cede2a2f-41a2-4748-9b12-c55c62f367ff EMERGENCY_CODES:M108,M112,M410
22:22:16.022 : N12 M220 S100*82
22:22:16.023 : X:0.00 Y:0.00 Z:0.00 E:0.00 Count X: 39087 Y:39087 Z:39087
22:22:16.023 : N13 M221 S100*82
22:22:16.023 : echo:DEBUG:INFO,ERRORS
22:22:16.023 : echo:Active Extruder: 0
22:22:16.023 : Begin file list
22:22:16.023 : N14 M111 S6*82
22:22:16.024 : N15 T0*14
22:22:16.027 : End file list
22:22:16.031 : echo:Hardcoded Default Settings Loaded
22:22:16.031 : echo:DEBUG:INFO,ERRORS
22:22:16.035 : echo:Active Extruder: 0
22:22:18.260 : N17 M500*16
22:22:19.381 : echo:Settings Stored (427 bytes)
22:22:31.757 : N22 G28*35
22:22:34.331 : X:0.00 Y:0.00 Z:317.06 E:0.00 Count X: 102585 Y:102625 Z:102500
22:22:34.332 : echo:endstops hit:  X:533.86 Y:534.07 Z:533.44
22:22:55.836 : N31 G28*33
22:22:56.825 : Error:checksum mismatch, Last Line: 31
22:22:56.828 : Resend: 32
22:23:12.512 : X:0.00 Y:0.00 Z:317.06 E:0.00 Count X: 148004 Y:149498 Z:205898
22:23:12.516 : Error:checksum mismatch, Last Line: 34
22:23:12.516 : Resend: 35
```

</details>
Result is same as Case1

A branch that it's used for Case7 and Case8:
https://github.com/esenapaj/Marlin/tree/testes5

Case9 and Case10 (same result):
<details>
<summary>

Newest RCBugFix, revert PR #4306 (ENDSTOPS_ONLY_FOR_HOMING is confusing), `#define ENDSTOPS_ONLY_FOR_HOMING` is enabled</summary>



```
22:38:47.325 : Printer reset detected - initalizing
22:38:47.325 : start
22:38:47.325 : echo: External Reset
22:38:47.325 : Marlin 1.1.0-RCBugFix
22:38:47.329 : echo: Last Updated: 2016-04-27 12:00 | Author: (Micromake)
22:38:47.329 : Compiled: Jul 19 2016
22:38:47.329 : echo: Free Memory: 2605  PlannerBufferBytes: 1408
22:38:47.333 : echo:V24 stored settings retrieved (427 bytes)
22:38:47.333 : echo:Steps per unit:
22:38:47.334 : echo:  M92 X200.00 Y200.00 Z200.00 E476.55
22:38:47.337 : echo:Maximum feedrates (mm/s):
22:38:47.337 : echo:  M203 X300.00 Y300.00 Z300.00 E300.00
22:38:47.337 : echo:Maximum Acceleration (mm/s2):
22:38:47.341 : echo:  M201 X3000 Y3000 Z3000 E9000
22:38:47.341 : echo:Accelerations: P=printing, R=retract and T=travel
22:38:47.345 : echo:  M204 P3000.00 R9000.00 T3000.00
22:38:47.354 : echo:Advanced variables: S=Min feedrate (mm/s), T=Min travel feedrate (mm/s), B=minimum segment time (ms), X=maximum XY jerk (mm/s),  Z=maximum Z jerk (mm/s),  E=maximum E jerk (mm/s)
22:38:47.354 : echo:  M205 S0.00 T0.00 B20000 X10.00 Z10.00 E5.00
22:38:47.355 : echo:Home offset (mm)
22:38:47.358 : echo:  M206 X0.00 Y0.00 Z0.00
22:38:47.358 : echo:Endstop adjustment (mm):
22:38:47.358 : echo:  M666 X0.00 Y0.00 Z0.00
22:38:47.362 : echo:Delta settings: L=diagonal_rod, R=radius, S=segments_per_second, ABC=diagonal_rod_trim_tower_[123]
22:38:47.366 : echo:  M665 L217.30 R95.00 S80.00 A0.00 B0.00 C0.00
22:38:47.366 : echo:Material heatup parameters:
22:38:47.366 : echo:  M145 S0 H200 B70 F255
22:38:47.370 : echo:  M145 S1 H240 B100 F255
22:38:47.370 : echo:PID settings:
22:38:47.370 : echo:  M301 P46.03 I6.24 D84.84 C100.00 L20
22:38:47.374 : echo:Retract: S=Length (mm) F:Speed (mm/m) Z: ZLift (mm)
22:38:47.375 : echo:  M207 S3.00 F2700.00 Z0.00
22:38:47.378 : echo:Recover: S=Extra length (mm) F:Speed (mm/m)
22:38:47.378 : echo:  M208 S0.00 F480.00
22:38:47.383 : echo:Auto-Retract: S=0 to disable, 1 to interpret extrude-only moves as retracts or recoveries
22:38:47.383 : echo:  M209 S0
22:38:47.383 : echo:Filament settings: Disabled
22:38:47.383 : echo:  M200 D1.75
22:38:47.386 : echo:  M200 D0
22:38:47.387 : echo:Z-Probe Offset (mm):
22:38:47.387 : echo:  M851 Z0.75
22:38:47.390 : current_position=(0.00, 0.00, 0.00) : sync_plan_position_delta
22:38:47.516 : N1 M110*34
22:38:47.516 : N2 M115*36
22:38:47.516 : N4 M114*35
22:38:47.552 : N5 M111 S6*98
22:38:47.557 : N6 T0*60
22:38:47.557 : N7 M20*22
22:38:47.561 : N8 M80*19
22:38:52.918 : echo:SD init fail
22:38:52.969 : FIRMWARE_NAME:Marlin 1.1.0-RCBugFix (Github) SOURCE_CODE_URL:https://github.com/MarlinFirmware/Marlin PROTOCOL_VERSION:1.0 MACHINE_TYPE:Micromake EXTRUDER_COUNT:1 UUID:cede2a2f-41a2-4748-9b12-c55c62f367ff EMERGENCY_CODES:M108,M112,M410
22:38:52.969 : N11 M220 S100*81
22:38:52.977 : N12 M221 S100*83
22:38:52.977 : X:0.00 Y:0.00 Z:0.00 E:0.00 Count X: 39087 Y:39087 Z:39087
22:38:52.977 : N13 M111 S6*85
22:38:52.978 : echo:DEBUG:INFO,ERRORS
22:38:52.978 : echo:Active Extruder: 0
22:38:52.978 : Begin file list
22:38:52.978 : End file list
22:38:52.978 : N14 T0*15
22:38:52.985 : echo:DEBUG:INFO,ERRORS
22:38:52.985 : echo:Active Extruder: 0
22:39:12.637 : N21 M502*23
22:39:12.640 : echo:Hardcoded Default Settings Loaded
22:39:14.293 : N23 M500*23
22:39:15.413 : echo:Settings Stored (427 bytes)
22:39:19.338 : N26 G28*39
22:39:21.811 : X:0.00 Y:0.00 Z:317.06 E:0.00 Count X: 102498 Y:102498 Z:102498
22:39:25.234 : N28 G28*41
22:39:26.259 : X:0.00 Y:0.00 Z:317.06 E:0.00 Count X: 102498 Y:102498 Z:102498
```

</details>

<details>
<summary>

Newest RCBugFix, revert PR #4306 (ENDSTOPS_ONLY_FOR_HOMING is confusing), `#define ENDSTOPS_ONLY_FOR_HOMING` is disabled</summary>



```
22:41:07.159 : Printer reset detected - initalizing
22:41:07.159 : start
22:41:07.159 : echo: External Reset
22:41:07.159 : Marlin 1.1.0-RCBugFix
22:41:07.163 : echo: Last Updated: 2016-04-27 12:00 | Author: (Micromake)
22:41:07.163 : Compiled: Jul 19 2016
22:41:07.163 : echo: Free Memory: 2604  PlannerBufferBytes: 1408
22:41:07.167 : echo:V24 stored settings retrieved (427 bytes)
22:41:07.167 : echo:Steps per unit:
22:41:07.171 : echo:  M92 X200.00 Y200.00 Z200.00 E476.55
22:41:07.171 : echo:Maximum feedrates (mm/s):
22:41:07.171 : echo:  M203 X300.00 Y300.00 Z300.00 E300.00
22:41:07.171 : echo:Maximum Acceleration (mm/s2):
22:41:07.175 : echo:  M201 X3000 Y3000 Z3000 E9000
22:41:07.179 : echo:Accelerations: P=printing, R=retract and T=travel
22:41:07.179 : echo:  M204 P3000.00 R9000.00 T3000.00
22:41:07.187 : echo:Advanced variables: S=Min feedrate (mm/s), T=Min travel feedrate (mm/s), B=minimum segment time (ms), X=maximum XY jerk (mm/s),  Z=maximum Z jerk (mm/s),  E=maximum E jerk (mm/s)
22:41:07.187 : echo:  M205 S0.00 T0.00 B20000 X10.00 Z10.00 E5.00
22:41:07.187 : echo:Home offset (mm)
22:41:07.191 : echo:  M206 X0.00 Y0.00 Z0.00
22:41:07.191 : echo:Endstop adjustment (mm):
22:41:07.191 : echo:  M666 X0.00 Y0.00 Z0.00
22:41:07.195 : echo:Delta settings: L=diagonal_rod, R=radius, S=segments_per_second, ABC=diagonal_rod_trim_tower_[123]
22:41:07.200 : echo:  M665 L217.30 R95.00 S80.00 A0.00 B0.00 C0.00
22:41:07.200 : echo:Material heatup parameters:
22:41:07.200 : echo:  M145 S0 H200 B70 F255
22:41:07.203 : echo:  M145 S1 H240 B100 F255
22:41:07.204 : echo:PID settings:
22:41:07.204 : echo:  M301 P46.03 I6.24 D84.84 C100.00 L20
22:41:07.208 : echo:Retract: S=Length (mm) F:Speed (mm/m) Z: ZLift (mm)
22:41:07.208 : echo:  M207 S3.00 F2700.00 Z0.00
22:41:07.212 : echo:Recover: S=Extra length (mm) F:Speed (mm/m)
22:41:07.212 : echo:  M208 S0.00 F480.00
22:41:07.216 : echo:Auto-Retract: S=0 to disable, 1 to interpret extrude-only moves as retracts or recoveries
22:41:07.216 : echo:  M209 S0
22:41:07.216 : echo:Filament settings: Disabled
22:41:07.220 : echo:  M200 D1.75
22:41:07.220 : echo:  M200 D0
22:41:07.220 : echo:Z-Probe Offset (mm):
22:41:07.220 : echo:  M851 Z0.75
22:41:07.223 : current_position=(0.00, 0.00, 0.00) : sync_plan_position_delta
22:41:07.317 : N1 M110*34
22:41:07.317 : N2 M115*36
22:41:07.317 : N4 M114*35
22:41:07.343 : N5 M111 S6*98
22:41:07.345 : N6 T0*60
22:41:07.345 : N7 M20*22
22:41:07.345 : N8 M80*19
22:41:12.753 : echo:SD init fail
22:41:12.803 : FIRMWARE_NAME:Marlin 1.1.0-RCBugFix (Github) SOURCE_CODE_URL:https://github.com/MarlinFirmware/Marlin PROTOCOL_VERSION:1.0 MACHINE_TYPE:Micromake EXTRUDER_COUNT:1 UUID:cede2a2f-41a2-4748-9b12-c55c62f367ff EMERGENCY_CODES:M108,M112,M410
22:41:12.803 : N11 M220 S100*81
22:41:12.812 : N12 M221 S100*83
22:41:12.812 : X:0.00 Y:0.00 Z:0.00 E:0.00 Count X: 39087 Y:39087 Z:39087
22:41:12.812 : N13 M111 S6*85
22:41:12.812 : echo:DEBUG:INFO,ERRORS
22:41:12.812 : echo:Active Extruder: 0
22:41:12.812 : Begin file list
22:41:12.813 : End file list
22:41:12.813 : N14 T0*15
22:41:12.819 : echo:DEBUG:INFO,ERRORS
22:41:12.819 : echo:Active Extruder: 0
22:41:14.007 : N16 M502*19
22:41:14.011 : echo:Hardcoded Default Settings Loaded
22:41:16.173 : N17 M500*16
22:41:17.291 : echo:Settings Stored (427 bytes)
22:41:18.351 : N19 G28*43
22:41:20.572 : X:0.00 Y:0.00 Z:317.06 E:0.00 Count X: 102498 Y:102498 Z:102498
22:41:22.982 : N22 G28*35
22:41:24.021 : X:0.00 Y:0.00 Z:317.06 E:0.00 Count X: 102498 Y:102498 Z:102498
```

</details>

Result is same as Case3 and Case4

A branch that it's used for Case9 and Case10:
https://github.com/esenapaj/Marlin/tree/testes4

AnHardt committed:
Introduce line_to_axis_pos(axis_codes axis, float where, float feed_rate = 0.0)

and use it in `homeaxis()` instead of `do_blocking_move_to_axis_pos()`.
`do_blocking_move_to_axis_pos` was wrong because it performed subdivided, delta-corrected moves for x- and y-axis.

The first common move for delta homing is like quick_home but for 3 towers.

Fix two warnings.

AnHardt commented:
Sorry for the mess.
Sleeping helped. Reverted #4338 and begun from scratch. Then i compared the new result with the old one - and .... -  the error was a unbelievable simple one - a simple missed `stepper.synchronize();` in the new function. If you thrust in its existence and remove it everywhere else, chaos is programmed. Sorry.

Let's have a new try.

The endstops have been under suspicion because without the synchronizes the G28-function is ready before the moves have finished. The endstops are switcht off again at the end of the function, but the moves in the planner buffer are still running.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -1613,9 +1613,20 @@ inline float set_homing_bump_feedrate(AxisEnum axis) {
   planner.buffer_line(current_position[X_AXIS], current_position[Y_AXIS], current_position[Z_AXIS], current_position[E_AXIS], MMM_TO_MMS(feedrate_mm_m), active_extruder);
 }
+
 inline void line_to_z(float zPosition) {
```
Note that `line_to_z(z)` does not update `current_position[Z_AXIS]`. It's called only in `run_z_probe` for the probing moves (where it doesn't need to update `current_position`). The `current_position` is set from `stepper.get_axis_position_mm(Z_AXIS)` at the end.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -1613,9 +1613,20 @@ inline float set_homing_bump_feedrate(AxisEnum axis) {
+  float old_feedrate_mm_m = feedrate_mm_m;
+  current_position[axis] = where;
+  feedrate_mm_m = (fr_mm_m != 0.0) ? fr_mm_m : homing_feedrate_mm_m[axis];
+  planner.buffer_line(current_position[X_AXIS], current_position[Y_AXIS], current_position[Z_AXIS], current_position[E_AXIS], MMM_TO_MMS(feedrate_mm_m), active_extruder);
```
Replacing this with `line_to_current_position()` makes the code a little smaller, but costs a function call.

thinkyhead commented:
Please let this fix many issues.

ghost commented:
About G28 of DELTA, I've confirmed that problem is solved by this PR.
But now, G29 of DELTA has been broken by PR #4356 (Unify run_z_probe)...

thinkyhead commented:
> G29 of DELTA has been broken

For f**\* sake. We need some way to have Marlin run in "simulation mode" where instead of controlling the hardware it just outputs what it _would_ do. Then people without deltas will be able to test their changes instead of submitting code full of naive assumptions.

jbrazio commented:
@daid's software simulator required the firmware to be built in a certain way in order to work.. what we would really need is mod it to allow .hex loading.. anyone with the skillz for the job ?

thinkyhead commented:
I'm thinking of something more along the lines of a simulation mode configuration option. All the low-level functions to control the hardware would be supplanted by functions that merely simulate everything. And then in this mode we would have a "monitor" that runs from `idle()`, watching the important variables and printing them to the console anytime they change.

thinkyhead commented:
I'm thinking of something more along the lines of a simulation mode configuration option. All the low-level functions to control the hardware would be supplanted by functions that merely simulate everything. And then in this mode we would have a "monitor" that runs from `idle()`, watching the important variables and printing them to the console anytime they change.

#4364 is the general idea.

jbrazio commented:
But that we can accomplish with a RAMPS board without anything connected to it, set `TEMP_SENSOR_0` to 998 and you're ready to rock. I would value more as a debugging tool a visual 3D representation of the kinematics running from an software emulated AVR.

-keep dreaming João-

thinkyhead commented:
If you can get MarlinSimulator to build on macOS (née OS X, née Mac OS X) you will be my hero.
